edb/schema/constraints.py

```diff
@@ -176,18 +176,27 @@ def get_name_impacting_ancestors(
             return [self.get_nearest_generic_parent(schema)]
 
     def get_constraint_origins(
-            self, schema: s_schema.Schema) -> List[Constraint]:
-        origins: List[Constraint] = []
+        self, schema: s_schema.Schema
+    ) -> List[Constraint]:
+        """
+        Find the transitive closure of all constraints that apply.
+        Return parents of these constraints.
+        """
+        # I'm not sure that this explanation is clear (or even entirely correct)
```

Comment:
I don't think this explanation is quite right. I think it would be worth having clear documentation here, because there is a subtle issue at play.

The purpose of this function is to find which types an exclusive constraint originate from. If we have `Baz <: Bar <: Foo`, and Foo declares some exclusive constraint, then when we do an insert into `Baz` we need to check all `Foo`s to make sure there is no conflict. (Which we do using triggers, and `get_constraint_origin` helps drive the generation of those.)

The subtleties here are:
 1. Because of multiple inheritance, we might have an exclusive constraint that originates from multiple unrelated ancestors
 2. Delegated exclusive constraints are declared on a parent, but their "origins" are their children. (So that `abstract type Named` or whatever can declare an exclusive name, but the names will be exclusive *within* each of the types that subtypes it.)

Reply:
I'll add these docs in a separate PR